res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,772 @@
+
+    var statusSysex = [0xF0, 0x00, 0x20, 0x7F, 0x03, 0x01, 0xF7];
+    midi.sendSysexMsg(statusSysex, statusSysex.length);
+
+    // enables 4 bottom pads fader cuts
+    var faderCutSysex = [0xF0, 0x00, 0x20, 0x7F, 0x03, 0xF7];
+    midi.sendSysexMsg(faderCutSysex, faderCutSysex.length);
+
+    // initialize leds for both decks
+    for (var i = 0; i < 2; i++) {
+        midi.sendShortMsg(0x90 + i, 0x00, 0x01); // play
+        midi.sendShortMsg(0x90 + i, 0x01, 0x01); // cue
+        midi.sendShortMsg(0x90 + i, 0x02, 0x01); // sync
+        midi.sendShortMsg(0x90 + i, 0x07, 0x7f); // scratch
+        midi.sendShortMsg(0x90 + i, 0x1B, 0x01); // pfl
+
+        midi.sendShortMsg(0x94 + i, 0x00, 0x7F); // hotcue
+        midi.sendShortMsg(0x94 + i, 0x0D, 0x01); // auto loop
+        midi.sendShortMsg(0x94 + i, 0x07, 0x01); // fader cuts
+        midi.sendShortMsg(0x94 + i, 0x0B, 0x01); // sample
+
+        midi.sendShortMsg(0x94 + i, 0x34, 0x01); // half
+        midi.sendShortMsg(0x94 + i, 0x35, 0x01); // double
+        midi.sendShortMsg(0x94 + i, 0x40, 0x01); // loop
+
+        // initialize leds for shifted buttons
+        midi.sendShortMsg(0x90 + i, 0x04, 0x01); // play
+        midi.sendShortMsg(0x90 + i, 0x05, 0x01); // cue
+        midi.sendShortMsg(0x90 + i, 0x03, 0x01); // sync
+        midi.sendShortMsg(0x94 + i, 0x0F, 0x01); // sample shifted
+        midi.sendShortMsg(0x94 + i, 0x02, 0x01); // beatjump
+        midi.sendShortMsg(0x90 + i, 0x08, 0x01); // bleep
+        midi.sendShortMsg(0x94 + i, 0x36, 0x01); // half
+        midi.sendShortMsg(0x94 + i, 0x37, 0x01); // double
+        midi.sendShortMsg(0x94 + i, 0x41, 0x01); // loop
+
+        // pads
+        for (var j = 0; j < 8; j++) {
+            midi.sendShortMsg(0x94 + i, 0x14 + j, 0x01);
+            midi.sendShortMsg(0x94 + i, 0x1C + j, 0x01); // shifted
+        }
+    }
+
+    midi.sendShortMsg(0x88, 0x09, 0x01); // tap led
+
+    // check if quantize is enabled
+    var quantizeEnabled = engine.getValue("[Channel1]", "quantize");
+
+    // effect leds
+    midi.sendShortMsg(0x88, 0x00, 0x01); // hpf
+    midi.sendShortMsg(0x88, 0x01, 0x01); // lpf
+    midi.sendShortMsg(0x88, 0x02, 0x01); // flanger
+    midi.sendShortMsg(0x89, 0x03, 0x01); // echo
+    midi.sendShortMsg(0x89, 0x04, quantizeEnabled ? 0x7F : 0x01); // reverb
+    midi.sendShortMsg(0x89, 0x05, 0x01); // phaser
+
+    // vumeters leds (off)
+    midi.sendShortMsg(0xB0, 0x1F, 0x00);
+    midi.sendShortMsg(0xB1, 0x1F, 0x00);
+
+    engine.makeConnection("[Channel1]", "VuMeter", MixtrackProFX.vuCallback);
+    engine.makeConnection("[Channel2]", "VuMeter", MixtrackProFX.vuCallback);
+};
+
+MixtrackProFX.shutdown = function() {
+    var shutdownSysex = [0xF0, 0x00, 0x20, 0x7F, 0x02, 0xF7];
+    midi.sendSysexMsg(shutdownSysex, shutdownSysex.length);
+};
+
+MixtrackProFX.EffectUnit = function(unitNumber) {
+    this.unitNumber = unitNumber;
+    this.group = "[EffectRack1_EffectUnit" + unitNumber + "]";
```

Comment:
```suggestion
MixtrackProFX.EffectUnit = function(deckNumber) {
    this.deckNumber = deckNumber;
    this.group = "[QuickEffectRack1_[Channel" + deckNumber + "]]";
```
This controller has no controls for the standard effect units, these central controls would be better used for QuickEffects.

Reply:
For Mixxx 2.4 the buttons should be mapped to using `load_preset` for the QuickEffects.

Reply:
I'm having trouble understanding how is this supposed to work. In the GUI I see that the filter knob is associated with QuickEffectRack1 super1. Now we also want to also use QuickEffectRack1 for FX. Won't those two clash? Filter knob needs to work independently of FX.

Another thing is that I don't see the effect selected in QuickEffectRack1 anywhere in the GUI (unless I'm missing something?), so it's very difficult to figure out which effect is currently selected without actually hearing it.

Reply:
Be is proposing that the central effect section is not controlling an EffectUnit but instead the QuickEffectRack. The QuickEffect can be configured in the `preferences > Equalizer` page right now. However with #2618 which is being planned to be part of 2.4 (possibly 2.5 if some things don't work out), controllers with central effect features like this one (or the traktor control S4 mk3) will be able to be mapped properly.

Reply:
Maybe I don't see the whole picture, but I still think that the current solution is better than using `QuickEffectRack`.

Now each deck can select a separate effect with buttons on the controller, control its params as well as dry/wet ratio, and the selected effect name is visible in the GUI.

With `QuickEffectRack` we effectively have only one effect, changing it requires going into settings dialog, and it's the same effect which is already controlled by the filter knob.

I understand that after some time we'll be able to properly map the effects (so the effect selection works the same as in Serato, VDJ, etc.), but I don't see how using `QuickEffectRack` is better for current Mixxx capabilities. Unless we want to target a future release where we'll be able to properly map it.

Reply:
Right, I think the current solution is fine for now. It would be great if you could update the mapping once the additional effects features have been merged. @Be-ing what do you think about h67ma's point?

Reply:
Considering the UX for changing the QuickEffect is quite clunky right now by requiring going into the preferences, I think it is okay for 2.3 to keep this mapped to the EffectUnits. I do think it should be changed for 2.4 and I think it would be good to add a TODO comment to the code mentioning this.

Reply:
I've added a TODO and subscribed to #2618, so I'll know when I can make the update.